### ISSUE TITLE
CUDA compiler doesn't support gcc versions greater than 8

### ISSUE THREAD

**Issue by Rampagy** [FoldingAtHome/fah-issues#1569]
Your issue may already be reported!
Please search on the [issue tracker](./) before creating one.

## Your Environment

* **F@H Software version**: 7.6.13
* **Operating System**: Fedora 32
* **Browser**: Not applicable, but Chrome

----
## Expected Behavior

The CUDA environment should be allowed with gcc versions > 8.  

----
## Current Behavior

The CUDA environment rolled out on Sept. 28th is unable to be initialized due to gcc version 10.2 being used on my computer.

----
## Possible Solution (Optional)

Allow CUDA environment to be built with gcc versions > 8.

----
## Steps To Reproduce

1.  I use a 1060 6Gb Nvidia GPU, but I think you can reproduce with any NVIDIA graphics card.
2.  Install gcc version 10.2.20200723 (Red Hat 10.2.1-1)
3.  Start FAH Client (7.6.13)
4.  Get OPENMM_22 GPU work unit

----
## Context

I would love to have my GPU contribute via the new CUDA drivers that were announced on September 28th, but it appears there is some error that won't allow it to happen if you have a gcc version >8.

Is this because I already have CUDA installed so FAHClient tried to use what was already there, but failed to realise it was a newer version and wouldn't compile?

----
[log.txt](https://github.com/FoldingAtHome/fah-issues/files/5302910/log.txt)

![Screenshot from 2020-09-29 22-09-09](https://user-images.githubusercontent.com/32419459/94638892-8982b680-02a0-11eb-8646-3c4b55c35e8c.png)
![Screenshot from 2020-09-29 21-49-33](https://user-images.githubusercontent.com/32419459/94638191-1167c100-029f-11eb-9de9-514808fa5472.png)

**Comment by PantherX** [comment]
If you mean that you have CUDA toolkit, then there's a known issue so uninstalling CUDA toolkit would potentially solve that. Development is aware of this issue and working with Nvidia to resolve this (it's a Linux/Windows issue).

I assume that you're running the latest Nvidia Drivers?

**Comment by Rampagy** [comment]
I have the CUDA toolkit installed from some of my AI projects.  Pytorch and tensorflow both need it.

I'm running NVIDIA driver 450.66.

![Screenshot from 2020-09-30 06-59-23](https://user-images.githubusercontent.com/32419459/94682488-a348eb80-02ea-11eb-94f9-5c55b2d856e3.png)

**Comment by Rampagy** [comment]
I tried uninstalling the CUDA toolkit, and now I can't fold on my GPU at all.  The 0x22 core keeps returning interrupted 106 (0x66).

[log.txt](https://github.com/FoldingAtHome/fah-issues/files/5305443/log.txt)

![Screenshot from 2020-09-30 07-29-06](https://user-images.githubusercontent.com/32419459/94685466-12c0da00-02ef-11eb-8974-604698bb0b80.png)

**Comment by jchodera** [comment]
@Rampagy: Thanks for reporting this!

Can you try a few things to help us debug?

**Run the core stand-alone:** First, go to the path where the core has been downloaded, something like `cores/cores.foldingathome.org/lin/64bit/22-0.0.13/Core_22.fah/`. Can you `ls -l` that directory and paste the output here? Then try running the core directly to make sure it doesn't segfault:
```bash
LD_LIBRARY_PATH="." ./FahCore_22 -info
```
Report the output here.

**Delete the core and have it re-download:** It's possible one of the dynamic libraries was accidentally deleted by an uninstall, in which case deleting the whole directory (**carefully** `rm -rf cores/cores.foldingathome.org/lin/64bit/22-0.0.13` or equivalent) will force the core to be re-downloaded and unpacked.

**Add the `-disable-cuda` flag:** Assuming that the previous step ran fine, can you try to configure the [`extra-core-args`](https://foldingathome.org/support/faq/installation-guides/configuration-guide/) for your GPU folding slot to add `-disable-cuda`, which should allow you to go back to OpenCL. Does this work? 

We have a few more things we can do to debug from here after we get this info back.

**Comment by Rampagy** [comment]
After this morning when I uninstalled CUDA and couldn't get my GPU to start folding again, I decided to reinstall the CUDA driver so that it would run until I got another response (it worked).  Now I am going through the same procedure I went through this morning and I decided to document every step.  

**Spoiler:** I don't get the same error anymore which is really frustrating....  That being said I get a new error!

**Step 1:** Uninstall CUDA.
[uninstalling_log.txt](https://github.com/FoldingAtHome/fah-issues/files/5308721/uninstalling_log.txt)

**Step 2:**  Check the core directory.
![cores_directory](https://user-images.githubusercontent.com/32419459/94750495-cce62f00-034b-11eb-8c6a-283946c1fb13.png)

**Step 3:** Run the core directly.
![RunningCoreDirectly](https://user-images.githubusercontent.com/32419459/94750536-e12a2c00-034b-11eb-94cd-f3a87113361d.png)

**Step 4:** Run the client with `-disable-cuda` flag.  It was at this point that I thought I knew what I was doing and decided to delete the GPUs.txt file because maybe it got corrupted.  That was not the case as you can see I still get the error after deleting GPUs.txt (so that it can re-downloaded).  Let me know what you find!

[log.txt](https://github.com/FoldingAtHome/fah-issues/files/5308736/log.txt)
[log-20200930-231133.txt](https://github.com/FoldingAtHome/fah-issues/files/5308744/log-20200930-231133.txt)
[log-20200930-231552.txt](https://github.com/FoldingAtHome/fah-issues/files/5308745/log-20200930-231552.txt)
[log-20200930-232723.txt](https://github.com/FoldingAtHome/fah-issues/files/5308746/log-20200930-232723.txt)
![WrongDriver](https://user-images.githubusercontent.com/32419459/94750780-80e7ba00-034c-11eb-8436-08510ee4bac0.png)

**Comment by Rampagy** [comment]
By now I'm sure you've checked out my Github and know I'm a tinkerer.  I just couldn't let this one go.  NVIDIA must have JUST released a new driver (450.80.02) at like 4pm today, because this was not up the last time I checked.

So I went about the process of updating my GPU driver.  Let me tell you it's a bitch to update your NVIDIA driver... Blacklisting nouveau via modprobe, while not being able to use Gnome display manager (gdm) because it keeps launching NVIDIA x-server, which can't be running while updating the NVIDIA driver... I now know why Linus Torvalds hates NVIDIA... https://youtu.be/IVpOyKCNZYw 

Rant over...

With the new NVIDIA driver (450.80.02) I am able to fold on my GPU and with CUDA cores nonetheless!  I appreciate all of your hard work and everything you've contributed to the F@H community!
[SUCCESS_LOG.txt](https://github.com/FoldingAtHome/fah-issues/files/5309170/SUCCESS_LOG.txt)
![NVIDIA_NEW_DRIVER](https://user-images.githubusercontent.com/32419459/94758977-49850780-0364-11eb-97ba-b921a125b3e5.png)

![SUCCESS_LOG](https://user-images.githubusercontent.com/32419459/94758821-e4c9ad00-0363-11eb-8d75-ea574a67d800.png)

One extra question:
Is there a way to set a max value for this when it's trying to get a new work unit?
![screen_mod](https://user-images.githubusercontent.com/32419459/94759338-40486a80-0365-11eb-9b18-318a5446f362.png)

**Comment by jchodera** [comment]
Interesting that `-disable-cuda` didn't work! It looks like the client wasn't finding a valid OpenCL installation, which is super weird if you had the CUDA drivers installed.

**Comment by Rampagy** [comment]
It was only when I uninstalled the CUDA drivers that OpenCL didn't work.  When I reinstalled CUDA OpenCL started working again.  My theory is uninstalling CUDA somehow uninstalled OpenCL (or broke it by uninstalling one of it's dependencies), then reinstalling the new graphics driver (450.80.02) reinstalled whatever broke it initially.

There were a whole bunch of things that were uninstalled with the CUDA driver because they were 'unused dependencies'.

[Uninstalling_log.txt](https://github.com/FoldingAtHome/fah-issues/files/5308721/uninstalling_log.txt)

**Comment by anand-bhat** [comment]
@Rampagy,
> Is there a way to set a max value for this when it's trying to get a new work unit?

No. This is determined and controlled by the client. The # Attempts and Next Attempt timestamp are populated when the client fails to get work and the client decides when it needs to try next. Prior to this version, the backoff was exponential (I think) capped at 6 hours. With the latest version (7.6.13), it still follows the same pattern but is set to a 1 hour maximum.

If you're however referring to the functionality where you request for a new work before the current work unit is done, there is a setting for that where you can state at what percentage of the current WU do you wish to ask for the new WU (the default in 99%).

**Comment by bb30994** [comment]
Are there entries in your path still pointing to the SDK (just a guess)?   Is a runtime compiler installed someplace that you didn't clean?  Maybe the uninstall didn't leave a clean system.  FAH delivers the pieces of CUDA that it needs along with FAHCore_22 so something unexpected is conflicting with it.

You'll still need the proprietary drivers from nV.

**Comment by Rampagy** [comment]
I think we are getting side tracked by my uninstall of CUDA and being unable to fold on my GPU (this is now fixed and no longer an issue).

Where are we at on the original issue I reported?

Is anyone able to reproduce the original issue where the CUDA compiler would fault out and FAHClient would fall back to OpenCL?

**Comment by bb30994** [comment]
The fault in a known problem and FAH is waiting on nV to discuss a final resolution. In the meantime the known work-around is to uninstall the SDK.
